### PR TITLE
Fix callSetsGenerator and variantsGenerator...

### PR DESCRIPTION
**Pull request by dcolligan**
... so that they send a null nextPageToken on
what should be their last response to the client.

Issue #267

### CODE REVIEW

File: ga4gh/backend.py [hunk 1]
```diff
@@ -222,13 +244,29 @@ def variantsGenerator(self, request):
...
+        """
+        Internal variants generator; returns (variant, nextPageToken) pairs
+        """
         variantSetIds = request.variantSetIds
         startVariantSetIndex = 0
         startPosition = request.start
         if request.pageToken is not None:
             startVariantSetIndex, startPosition = self.parsePageToken(
                 request.pageToken, 2)
-        for variantSetIndex in range(startVariantSetIndex, len(variantSetIds)):
```
Comment (jeromekelleher) [inline ga4gh/backend.py]: Why the line break here after range?

**Comment by jeromekelleher** [comment]
I think I'd rather deal with this at the source. It'll be confusing if we have some of our generators having one behaviour (correct null pageToken for the last value) and others having different behaviour which need a wrapper. We need to update `variantsGenerator` so that it deals with  variants overlapping the page boundary correctly, and we might as well use the code we're developing for reads to do this. We should just fix callSetsGenerator.

**Comment by dcolligan** [comment]
How is that confusing?  I think the code is pretty easy to follow.

Plus, this gives us a general solution to a problem that more than one of the generators have now, and more are likely to have in the future.  This is way better than having every generator implementing look-ahead logic.

**Comment by jeromekelleher** [comment]
I think #295 fixes the callSets stuff (using the general code for lists of objects), and we will need to review variants in any case (hopefully reusing the logic we have for reads).

**Comment by jeromekelleher** [comment]
I think this has been superseded now --- shall we close this PR @dcolligan, or are there bits you'd like to bring forward?

**Comment by dcolligan** [comment]
I do like the idea of having a general construct for implementing look-ahead to ensure that we send a null nextPageToken with the last data object yielded.  I think that this would still be worth implementing.  Do you agree?  It looks like we would only use this in `variantsGenerator` and `readsGenerator` right now (the only generators not using `_topLevelObjectGenerator`).

**Comment by jeromekelleher** [comment]
Well, variantsGenerator is going to need to change to use the same algorithm as readsGenerator because the same issues with overlapping intervals can occur (albeit with a much lower probability). So,  we'll be using the same code for both of these in any case. 

I agree that implementing read-ahead differently for the different iterators is a bad idea. I would suggest that the next time we look at the paging code (which we're definitely doing to need to do when we tackle datasets) we incorporate something like you've defined here.

**Comment by dcolligan** [comment]
Ok I'll close the PR but keep the branch around

**Pull request closed without merging**